rfcs/0027-frontier-synchronization.md

```diff
@@ -0,0 +1,188 @@
+3. Once all the deferreds in #1 resolve, transmute the write action returned by the write deferred into a list of diffs
+4. Apply all the diffs from #2 to the full frontier
+5. Write the diffs from #2 into the persistent frontier sync diff buffer asynchronously
+6. Update each frontier extension with the diffs from #2
+7. Notify frontier extension subscribers of any frontier extension changes
+
+#### Full/Partial Breadcrumbs + Any Refs
+
+The new synchronization cycle addresses synchronization issues between the state of the full frontier datastructure, the persistent root database, and the transition frontier extensions, but there is still a remaining synchronization issue regarding breadcrumbs. Reading breadcrumbs from the transition frontier is synchronized within the new synchronization cycle, but staged ledger mask accesses cannot be sanely synchonized within that same cycle. In order to address this issue, breadcrumbs are now separated into two types: full breadcrumbs (which contain staged ledger's with masks attached to the frontier's persistent root, directly or indirectly) and partial breadcrumbs (which only contain the staged ledger's scan state and not the full data structure with the mask). Reading a breadcrumb from the frontier will return neither a full or partial breadcrumb, but instead will return a `Breadcrumb.Any.Ref.t`, which is a reference to either a full or partial breadcrumb. This value can be safely passed back from outside the `'a Transition_frontier.Read.t` monad and the "staged state" (either the staged ledger or scan state depending on the status of the breadcrumb) can be safely queried at an arbitrary time in the future. The downside of this technique is that only immediate (and synchronous) reads from the breadcrumb's staged ledger mask are safe under this interface, disallowing certain actions. For instance, under this system, it is not safe to use a breacrumb's staged ledger mask as a target for the syncable ledger (which probably isn't safe anyway for other reasons).
+
+#### Detached Mask State Management
+
+In the current system, when a mask is destroyed, all of it's children masks are also destroyed. Masks will now additionally provide the ability to attach "destruction hooks". This will be used to cleanup full breadcrumbs and downgrade them to partial breadcrumbs if the breadcrumb's parent mask chain becomes inaccessible. This allows code outside of synchronized frontier reads to continue to query breadcrumbs from the frontier and build successor breadcrumbs off of them safely. This is important in the context of catchup since new subtrees of breadcrumbs are built asynchronously to the frontier being updated. If a subtree is built off of a breadcrumb which is later removed from the frontier (thus having it's mask destroyed), then that subtree will subsequently be marked as partial and any masks in it will be destroyed and inaccessible.
```

Comment:
The only time one of a catchup job's ancestor masks will be deleted without the reparenting thing happening is if the root is moved forward such that the ancestor is no longer in the frontier, correct? In that case the job should be cancelled anyway, since the chain we're catching up to lost and the point of divergence is older than the point of finality.

Reply:
Yes, it's true that this is the only time a catchup job's ancestor masks will be deleted. And I agree, the catchup job should be canceled when this happens. However, we still need this system for downgrading the breadcrumbs and taking the masks out of scope as it provides a simple way for catchup jobs to notice that the subtree is now invalid while also solving synchronization issues in between the completion of catchup and the inclusion of the new breadcrumb subtree into the frontier.